akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +697,385 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
     (logic, Source.fromGraph(source))
   }
 }
+
+/**
+ * FIXME write docs
+ *
+ * A BroadcastHub is a special streaming hub that is able to broadcast streamed elements to a dynamic set of consumers.
+ * It consists of two parts, a [[Sink]] and a [[Source]]. The [[Sink]] broadcasts elements from a producer to the
+ * actually live consumers it has. Once the producer has been materialized, the [[Sink]] it feeds into returns a
+ * materialized value which is the corresponding [[Source]]. This [[Source]] can be materialized an arbitrary number
+ * of times, where each of the new materializations will receive their elements from the original [[Sink]].
+ */
+object PartitionHub {
+
+  /**
+   * FIXME write docs
+   *
+   * Creates a [[Sink]] that receives elements from its upstream producer and broadcasts them to a dynamic set
+   * of consumers. After the [[Sink]] returned by this method is materialized, it returns a [[Source]] as materialized
+   * value. This [[Source]] can be materialized an arbitrary number of times and each materialization will receive the
+   * broadcast elements form the ofiginal [[Sink]].
+   *
+   * Every new materialization of the [[Sink]] results in a new, independent hub, which materializes to its own
+   * [[Source]] for consuming the [[Sink]] of that materialization.
+   *
+   * If the original [[Sink]] is failed, then the failure is immediately propagated to all of its materialized
+   * [[Source]]s (possibly jumping over already buffered elements). If the original [[Sink]] is completed, then
+   * all corresponding [[Source]]s are completed. Both failure and normal completion is "remembered" and later
+   * materializations of the [[Source]] will see the same (failure or completion) state. [[Source]]s that are
+   * cancelled are simply removed from the dynamic set of consumers.
+   *
+   * @param bufferSize Buffer size used by the producer. Gives an upper bound on how "far" from each other two
+   *                   concurrent consumers can be in terms of element. If this buffer is full, the producer
+   *                   is backpressured. Must be a power of two and less than 4096.
+   */
+  def sink[T](partitioner: T ⇒ Int, startAfterNbrOfStreams: Int, bufferSize: Int = 256): Sink[T, Source[T, NotUsed]] =
```

Comment:
`(T, Int) => Int` works for the hashCode modulo use case, but I think there can be other use cases.
1) sticky session, like Johan mention, e.g. route all elements for user "Bob" to a specific stream, until that stream disappears and then pick another
2) round robin for each element, also requires state (the counter) in the function as another delicate

In the implementation each stream has a `Long` id (it's just a counter). I'm thinking of exposing those as `Array[Long] => Long`. Then you know the size, and could keep track of a sticky session mapping. To support state it has to be `() => (Array[Long] => Long)`, which is always a bit mind boggling.

We could definitely have a simple and a more powerful API. We could also provide pre-packaged functions, such as for the round-robin case.